### PR TITLE
CIRC-920: Use effective permanent location of an item when determining owner.

### PR DESCRIPTION
**Pull request by bohdan-suprun**
https://issues.folio.org/browse/CIRC-920 - Wrong location used to determine fee/fine owner for items declared lost.

## Purpose
Item "effective" permanent location should be used to find fee/fine owner for a lost fee.

The algorithm:
* Use Item location: Permanent
* If Item location: Permanent does not exist, use Holdings location: Permanent
* If Holdings location: Permanent does not exist, use Effective location for item

Note: the last is not possible, because holdings permanent location is a required property.

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/circulation/domain/Item.java [hunk 1]
```diff
@@ -350,6 +351,13 @@ public boolean doesNotHaveHolding() {
     return holdingRepresentation == null;
   }
 
+  public String getPermanentLocationId() {
+    final String itemLocation = getProperty(itemRepresentation, PERMANENT_LOCATION_ID);
+    final String holdingsLocation = getProperty(holdingRepresentation, PERMANENT_LOCATION_ID);
+
+    return StringUtils.firstNonBlank(itemLocation, holdingsLocation);
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/Item.java:358]: Personally, I might be tempted to statically import `firstNonBlank`
```suggestion
    return firstNonBlank(itemLocation, holdingsLocation);
```

File: src/main/java/org/folio/circulation/infrastructure/storage/inventory/LocationRepository.java [hunk 2]
```diff
@@ -60,15 +60,23 @@ public static LocationRepository using(Clients clients) {
...
-      .fetch(item.getLocationId())
-      .thenApply(r -> r.map(Location::from))
+    return fetchLocationById(item.getLocationId())
       .thenCompose(r -> r.after(this::loadLibrary))
       .thenCompose(r -> r.after(this::loadCampus))
       .thenCompose(r -> r.after(this::loadInstitution));
   }
 
+  public CompletableFuture<Result<Location>> fetchLocationById(String id) {
+    if(StringUtils.isBlank(id)) {
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/infrastructure/storage/inventory/LocationRepository.java:70]:
```suggestion
    if (StringUtils.isBlank(id)) {
```

File: src/main/java/org/folio/circulation/infrastructure/storage/inventory/LocationRepository.java [hunk 3]
```diff
@@ -60,15 +60,23 @@ public static LocationRepository using(Clients clients) {
...
+      return ofAsync(() -> null);
+    }
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/infrastructure/storage/inventory/LocationRepository.java:72]: Is it possible to write a unit test to check this guard clause?

File: src/main/java/org/folio/circulation/services/LostItemFeeChargingService.java [hunk 4]
```diff
@@ -142,9 +142,12 @@ private UUID getOwnerServicePoint(Loan loan) {
...
 
   private CompletableFuture<Result<ReferenceDataContext>> fetchFeeFineOwner(ReferenceDataContext referenceData) {
-    final String servicePointId = getOwnerServicePoint(referenceData.loan).toString();
+    final String permanentLocationId = referenceData.loan.getItem().getPermanentLocationId();
 
-    return feeFineOwnerRepository.findOwnerForServicePoint(servicePointId)
+    return locationRepository.fetchLocationById(permanentLocationId)
+      .thenApply(r -> r.map(Location::getPrimaryServicePointId))
+      .thenApply(r -> r.map(UUID::toString))
+      .thenCompose(r -> r.after(feeFineOwnerRepository::findOwnerForServicePoint))
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/services/LostItemFeeChargingService.java:150]: Could this mapping from UUID to String either be avoided or moved to an overload in the fee fine owner repository?

File: src/test/java/org/folio/circulation/domain/ItemTest.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+import java.util.UUID;
+
+import org.junit.Test;
+
+import api.support.builders.HoldingBuilder;
+import api.support.builders.ItemBuilder;
+import lombok.val;
+
+public class ItemTest {
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/circulation/domain/ItemTest.java:15]: Could it worth making this class specific to location tests e.g. `ItemLocationTests` to keep the scope of the tests small?

**Review by marcjohnson-kint — COMMENTED** [review]
Thank you for introducing unit tests for determining the effective permanent location

File: src/main/java/org/folio/circulation/infrastructure/storage/inventory/LocationRepository.java (hunk 3 above)
Comment (bohdan-suprun) [reply src/main/java/org/folio/circulation/infrastructure/storage/inventory/LocationRepository.java]: Sure. added.

File: src/main/java/org/folio/circulation/services/LostItemFeeChargingService.java (hunk 4 above)
Comment (bohdan-suprun) [reply src/main/java/org/folio/circulation/services/LostItemFeeChargingService.java]: Added an override method.

File: src/test/java/org/folio/circulation/domain/ItemTest.java (hunk 5 above)
Comment (bohdan-suprun) [reply src/test/java/org/folio/circulation/domain/ItemTest.java]: Renamed to `ItemPermanentLocationTest`

**Review by marcjohnson-kint — APPROVED** [review]

**Pull request merged**